rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,261 @@
+  of words related by form or meaning.
+- _morphology (linguistics)_: the study of words, how they are formed, 
+  and their relationship to other words in the same language. 
+- _fulltext search index_: the result of lexical and morphological 
+  analysis (stemming) of a set of text documents.  It provides frequency 
+  and location for the language-specific stems found in the text documents 
+  being indexed. 
+- _ranking algorithm_: "Ranking attempts to measure how relevant documents 
+  are to a particular query, so that when there are many matches the most 
+  relevant ones can be shown first." [- Postgres Documentation](https://www.postgresql.org/docs/11/textsearch-controls.html#TEXTSEARCH-RANKING-SEARCH-RESULTS)
+  
+  **Algorithms**:
+  - _standard ranking_: ranking based on the number of matching lexemes.     
+  - _cover density ranking_: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.            
+
+## Detailed Design
+
+### Subgraph Schema
+
+Part of the power of the fulltext search API is the flexibility, so 
+it is important to expose a simple interface to facilitate useful applications 
+of the index and aim to reduce the need to create new subgraphs for the 
+express purpose of updating fulltext search fields. 
+
+For each fulltext search API a subgraph developer must be able to specify:
+    1. a language, 
+    2. a set of text document fields to include,
+    3. relative weighting for each field,
+    4. a choice of ranking algorithm for sorting query result items.
+
+The proposed process of adding one or more fulltext search API involves
+adding one or more fulltext directive to the `_Schema_` type in the
+subgraph's GraphQL schema. Each fulltext definition will have four
+required top level parameters: `name`, `language`, `algorithm`, and
+`include`. The fulltext search definitions will be used to generate
+query fields on the GraphQL schema that will be exposed to the end user.
+
+Enabling fulltext search across entities will be a powerful abstraction 
+that allows users to search across all relevant entities in one query. Such 
+a search will by definition have polymorphic results. To address this, a 
+union type will be generated in the schema for the fulltext search results. 
+
+Validation of the fulltext definition will ensure that all fields referenced 
+in the directive are valid String type fields. With subgraph composition 
+it will be possible to easily create new subgraphs that add specific fulltext 
+search capabilities to an existing subgraph. 
+
+Example fulltext search definition:
+
+```graphql
+type _Schema_ 
+  @fulltext(
+    name: "media"
+    ...
+  )
+  @fulltext(
+    name: "search",
+    language: "english",
```

Comment:
We should use [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) codes for the language (like `en` for English, `de` for German etc.) Everything else on the web uses them, too.

Reply:
I'm open to that though I think the full names are a bit more clear because they are self-describing.

Reply:
They are contrary to anywhere else where systems talk about human languages (e.g., HTTP, localization libraries, etc.) which all use the ISO codes. It's also not quite true that `French` is called `French` or `Spanish` `Spanish` ;)

Reply:
One counter example is Postgres which doesn't use the ISO codes for its fulltext search language configuration values, but I like the idea of using a standard that isn't english centric.  
I'll update the RFC to use `ISO 639-1` codes for the language value.